indexer/SymbolFormatter.cc

```diff
@@ -40,44 +41,197 @@ std::string_view SymbolFormatter::getMacroSymbol(clang::SourceLocation defLoc) {
+  // give correct results. In particular, the canonical decl may
+  // change depending on include order. For example, if you have:
+  //   void f(int x);     // In A.h
+  //   void f(int x = 0); // In B.h
+  // Then depending on #include order of A.h and B.h, the canonical decl
+  // will be different, since it is most commonly determined based on which
+  // one appears first.
+  //
+  // However, this doesn't matter because the symbol name of the function
+  // can only depend on:
+  // - The path of the main TU (if there is an anonymous namespace
+  //   in the decl context chain)
+  // - Names of namespaces.
+  // It cannot depend on the path of the header itself (header paths affect
+  // symbol names for anonymous types, but methods in different anonymous
+  // type declarations would never be mapped to the same symbol anyways),
+  // so even if the exact canonical decl is different, the symbol name
+  // will be the same exactly when required.
+  decl = decl->getCanonicalDecl();
+  auto it = this->declBasedCache.find(decl);
   if (it != this->declBasedCache.end()) {
     return std::string_view(it->second);
   }
-  if (namespaceDecl->isAnonymousNamespace()) {
-    auto mainFileId = this->sourceManager.getMainFileID();
-    ENFORCE(mainFileId.isValid());
-    auto path = this->getCanonicalPath(mainFileId);
-    if (!path.has_value()) {
-      // Strictly speaking, this will be suboptimal in the following case:
-      // - header 1: in source tree (has canonical path), uses namespace {..}
-      // - header 2: in source tree (has canonical path), uses namespace {..}
-      // - generated C++ file: in build tree only (no canonical path), and
-      //   includes header 1 and 2.
-      // We will not emit a symbol that connects the anonymous namespace
-      // in header 1 and header 2. However, that is OK, because it is unclear
-      // how to handle this case anyways, and anonymous namespaces are
-      // rarely (if ever) used in headers.
-      return {};
-    }
-    fmt::format_to(std::back_inserter(this->scratchBuffer), "$ANON/{}\0",
-                   path->asStringView());
-  } else {
-    llvm::raw_string_ostream os(this->scratchBuffer);
-    namespaceDecl->printQualifiedName(os);
-    absl::StrReplaceAll({{"::", "/"}}, &this->scratchBuffer);
+  auto symbol = getSymbol();
+  if (symbol.empty()) {
+    return "";
   }
-  std::string out{
-      fmt::format("c . todo-pkg todo-version {}/", this->scratchBuffer)};
-  this->scratchBuffer.clear();
   auto [newIt, inserted] =
-      this->declBasedCache.insert({namespaceDecl, std::move(out)});
+      this->declBasedCache.insert({decl, std::move(symbol)});
   ENFORCE(inserted);
   return std::string_view(newIt->second);
 }
 
+std::optional<std::string_view>
+SymbolFormatter::getContextSymbol(const clang::DeclContext *declContext) {
+  if (auto namespaceDecl = llvm::dyn_cast<clang::NamespaceDecl>(declContext)) {
+    return this->getNamespaceSymbol(namespaceDecl).value_or("");
```

Comment:
The `value_or("")` would've been triggered when we try to create the name for an anonymous namespace inside a generated TU main file (that's one of the kinds of files which doesn't have a canonical path currently).

However, this usage is wrong because it was added when `getContextSymbol` was returning a `string_view`. However, this function returns an optional now, so the nullopt case should be used instead.

Fixed in [`1a954d1` (#104)](https://github.com/sourcegraph/scip-clang/pull/104/commits/1a954d160b8ddc66fdba102b435a929cb1804335)